### ISSUE TITLE
ReinforcementLearner does not fetch/store candles in dry-run

### ISSUE THREAD

**Issue by aredhil89** [freqtrade/freqtrade#11710]
### Describe your environment
- OS: Docker on Windows 10 (WSL2 backend)
- Freqtrade Version: 2025.4 (latest `develop`)
- Python Version: 3.12.10
- Running Mode: `dry_run` via Docker
- Strategy: `HybridAI_AI` (custom, inherit from FreqaiStrategy)
- AI: FreqAI enabled with `ReinforcementLearner` (PPO)
- Database: PostgreSQL enabled
- Candle storage: Enabled (FreqAI training mode)

---

### Your question / problem

I am facing an issue where **live candles are not being fetched** during dry-run trading mode. As a result:
- FreqAI cannot process or train on live data.
- No new candles appear in PostgreSQL `freqai_ohlcv_data`.
- `populate_indicators()` runs, but logs confirm that candle data is not updating.
- Strategy uses `super().populate_indicators()` and calls `self.freqai.start()` inside it.

I want to confirm:
- What **minimum requirements or rules** must be followed to ensure FreqAI is able to fetch live candles in dry-run?
- Does the strategy file need to implement anything specific beyond `super().populate_indicators()`?
- Are there **best practices or guidelines** to ensure FreqAI’s `DataProvider` is properly triggered during dry-run?

---

### Steps I have tried
- Verified `freqai_interface` is **not None** inside `populate_indicators()`.
- Checked FreqAI logs — shows “Start() executed” but no new candles processed.
- PostgreSQL is working and storing other tables (trades, etc).
- Config contains:
  ```json
  "freqai": {
    "enabled": true,
    "training_enabled": true,
    "model_save_enabled": true,
    "data_split_count": 2,
    ...
  }

**Comment by xmatthias** [comment]
To even start to look at this - please make sure to update your issue to have proper formatting.
You'r (CLEARLY AI generated issue) is pure html - formatting of all files is wrong - and honestly - i don't care fixing your broken html to make it readable / understandable.

It's most likely a configuration issue (or an issue with your strategy) - BUT - that's not really visible due to the broken formatting.

_note_: this issue will be closed within a day as "ai generated crap" if we don't get the formatting fixed - as in it's current state - it's not something i'll look at.

**Comment by xmatthias** [comment]
Well this didn't improve things. 
We got no logs, no strategy to go on with ...


From a quick look at what remains - there is no `freqai_interface` in freqtrade - which is a clear sign of a AI hallucinated strategy.

Please start from the [documentation example](https://www.freqtrade.io/en/stable/freqai/) and compare what's wrong with your strategy.

Please understand that we're not here to correct your AI's mistakes.
We're humans trying to help humans, but assume that said humans will take the time to read the documentation we take the time to write.

**Comment by aredhil89** [comment]
Thank you very much for your feedback and for pointing me in the right direction.

I truly appreciate all the effort, time, and documentation the Freqtrade team has provided. I’ll take a step back and review my strategy carefully based on the official documentation, and I’ll do my best to correct the issues on my own.

If I still encounter problems, I’ll try to ask here with a clearer explanation and hope to get a bit of guidance.
Thanks again for your patience and support 🙏

**Comment by aredhil89** [comment]
here i attach my strategies, fyi im still new and still working on it almost 2 month already,im very im very eager to learn.but im not sure if my strategies is correct or not.

``` python
import logging
import os
from datetime import datetime
from typing import Dict, Any
import talib.abstract as ta
from freqtrade.persistence import Trade
from pandas import DataFrame
import pandas as pd

import psycopg2
from freqtrade.freqai.freqai_strategy import FreqaiStrategy

from user_data.ai_modules.sentiment_analyzer import SentimentAnalyzer
from user_data.ai_modules.ai_decision_engine import AIDecisionEngine

logger = logging.getLogger(__name__)


class HybridAI_AI(FreqaiStrategy):
    INTERFACE_VERSION = 3
    timeframe = '1m'
    startup_candle_count: int = 150
    process_only_new_candles = True
    use_exit_signal = True
    exit_profit_only = False
    ignore_roi_if_entry_signal = False
    disable_dataframe_checks = False

    minimal_roi = {"0": 0.1, "30": 0.05, "60": 0.01, "120": 0}
    stoploss = -0.05
    trailing_stop = True
    trailing_stop_positive = 0.08
    trailing_stop_positive_offset = 0.14
    trailing_only_offset_is_reached = True

    order_types = {
        'entry': 'limit',
        'exit': 'limit',
        'stoploss': 'market',
        'emergency_exit': 'market',
        'force_entry': 'market',
        'force_exit': 'market',
        'stoploss_on_exchange': False
    }

    order_time_in_force = {
        'entry': 'GTC',
        'exit': 'GTC',
        'stoploss': 'GTC'
    }

    stake_currency = "USDT"
    stake_amount = 100
    max_open_trades = 2

    def __init__(self, config: Dict[str, Any]) -> None:
        super().__init__(config)
        self.sentiment_analyzer = SentimentAnalyzer(config)
        self.ai_engine = AIDecisionEngine(sentiment_analyzer=self.sentiment_analyzer)

        addon = config.get("addon_settings", {})
        self.ai_buy_threshold = float(addon.get("ai_buy_threshold", 0.6))
        self.ai_sell_threshold = float(addon.get("ai_sell_threshold", 0.4))
        logger.info(f"📈 AI thresholds: BUY > {self.ai_buy_threshold}, SELL < {self.ai_sell_threshold}")

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        pair = metadata.get("pair", "?")
        logger.info(f"[{pair}] 📍 Entering populate_indicators() with {len(dataframe)} candles")

        # Feather file check
        try:
            path = self.config.get("data_directory", "user_data/data")
            file_path = f"{path}/binance/{pair.replace('/', '')}-{self.timeframe}.feather"
            if os.path.exists(file_path):
                ts = os.path.getmtime(file_path)
                logger.info(f"[{pair}] 🕒 Feather last updated: {datetime.fromtimestamp(ts)}")
            else:
                logger.warning(f"[{pair}] ⚠️ Feather file not found: {file_path}")
        except Exception as e:
            logger.warning(f"[{pair}] ⚠️ Feather check error: {e}")

        try:
            dataframe = super().populate_indicators(dataframe, metadata)
            logger.info(f"[{pair}] ✅ super().populate_indicators() called.")

            dataframe = self.freqai.start(dataframe, metadata, self)
            logger.info(f"[{pair}] ✅ FreqAI .start() executed.")

            if self.freqai.freqai_model and hasattr(self.freqai.freqai_model, "model"):
                logger.info(f"[{pair}] ✅ Model loaded: {type(self.freqai.freqai_model.model).__name__}")
            else:
                logger.warning(f"[{pair}] ⚠️ FreqAI model not attached.")

            if not dataframe.empty and "date" in dataframe.columns:
                logger.info(f"[{pair}] 🟢 First candle: {dataframe.iloc[0]['date']}")
                logger.info(f"[{pair}] 🔴 Last candle: {dataframe.iloc[-1]['date']}")

                now = datetime.utcnow()
                last_ts = pd.to_datetime(dataframe.iloc[-1]["date"])
                diff = (now - last_ts).total_seconds()
                logger.info(f"[{pair}] ⏱️ Last candle age: {diff:.1f} sec ago")

                if 30 < diff < 120:
                    logger.info(f"[{pair}] ✅ Likely just closed and written to DB/feather")
                elif diff <= 30:
                    logger.info(f"[{pair}] 🔄 Live candle (not yet closed)")
                else:
                    logger.warning(f"[{pair}] ⚠️ Candle too old — check data feed?")
            else:
                logger.warning(f"[{pair}] 🔴 DataFrame empty or missing 'date' column.")

            # ✅ Check PostgreSQL DB for candle count
            try:
                conn = psycopg2.connect(
                    dbname=os.getenv("POSTGRES_DB", "freqtrade"),
                    user=os.getenv("POSTGRES_USER", "freqtrade"),
                    password=os.getenv("POSTGRES_PASSWORD", "freqtrade"),
                    host=os.getenv("POSTGRES_HOST", "localhost"),
                    port=os.getenv("POSTGRES_PORT", "5432")
                )
                cursor = conn.cursor()
                table_name = "freqai_ohlcv_data"
                q = f"SELECT COUNT(*) FROM {table_name} WHERE pair = %s AND timeframe = %s"
                cursor.execute(q, (pair, self.timeframe))
                count = cursor.fetchone()[0]
                logger.info(f"[{pair}] 🗃️ Total rows in DB: {count}")
                cursor.close()
                conn.close()
            except Exception as e:
                logger.warning(f"[{pair}] ⚠️ DB check failed: {e}")

        except Exception as e:
            logger.error(f"[{pair}] ❌ populate_indicators error: {e}")

        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        pair = metadata.get("pair", "?")
        dataframe['enter_long'] = 0
        dataframe['enter_short'] = 0
        try:
            sentiment = self.sentiment_analyzer.get_sentiment_score(pair)
            prediction_col = f"freqai_prediction_{pair.replace('/', '')}"

            if prediction_col in dataframe.columns:
                pred = dataframe[prediction_col]
                adjusted = pred + (sentiment * 0.25)
                dataframe.loc[adjusted > self.ai_buy_threshold, 'enter_long'] = 1
                dataframe.loc[adjusted < self.ai_sell_threshold, 'enter_short'] = 1
                logger.debug(f"[{pair}] ✅ Entry triggered with sentiment {sentiment}")
            else:
                logger.warning(f"[{pair}] ❌ Missing prediction column: {prediction_col}")
        except Exception as e:
            logger.warning(f"[{pair}] ⚠️ Entry trend error: {e}")
        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe['exit_long'] = (dataframe["&-action"] == 2).astype(int)
        dataframe['exit_short'] = (dataframe["&-action"] == 4).astype(int)
        return dataframe

    def set_freqai_targets(self, dataframe: DataFrame, metadata: dict, **kwargs) -> DataFrame:
        label_period = self.freqai_info["feature_parameters"]["label_period_candles"]
        dataframe["&-s_close"] = (
            dataframe["close"].shift(-label_period).rolling(label_period).mean() / dataframe["close"] - 1
        )
        return dataframe

    def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime,
                        current_rate: float, current_profit: float, **kwargs) -> float:
        if current_profit > 0.04:
            return max(min(current_profit / 2, 0.05), 0.025)
        return 1
```

**Comment by xmatthias** [comment]
Honestly? 
Scrap it, start from scratch using the documentation and  the Sample strategy we provide as starting point.
There's too much wrong with this for me to bother trying to help you fix it.

Don't use AI to generate it - but try to actually **read the documentation** instead.


The errors start at the imports - this file/strategy will NOT even start - as imports will fail unless it's a modified version of freqtrade. if it is - then it's a modified version of freqtrade - which for obvious reasons, we neither care about - and will NOT provide you assistance with.

looking forward (i made the mistake to start at the bottom - i'd have stopped here if i hadn't typed the below already ...) 

There's a (completely pointless) check on a table that doesn't exist (candle data is not stored in the database).

There's assumptions on "checking for feather file" - which (depending on the market and configuration) checks in the completely wrong location. While the filename is correct - it's not said that freqtrade is using feather data (this can be configured).
